test/proxy_test.go

```diff
@@ -0,0 +1,666 @@
+// connects to the backend server.
+func (s) TestGrpcDialWithProxy(t *testing.T) {
+	backendAddr := createAndStartBackendServer(t)
+	proxyListener, errCh, doneCh, _ := setupProxy(t, backendAddr, false, func(req *http.Request) error {
+		if req.Method != http.MethodConnect {
+			return fmt.Errorf("unexpected Method %q, want %q", req.Method, http.MethodConnect)
+		}
+		if req.URL.Host != unresolvedTargetURI {
+			return fmt.Errorf("unexpected URL.Host %q, want %q", req.URL.Host, unresolvedTargetURI)
+		}
+		return nil
+	})
+
+	hpfe := func(req *http.Request) (*url.URL, error) {
+		if req.URL.Host == unresolvedTargetURI {
+			return &url.URL{
+				Scheme: "https",
+				Host:   proxyListener.Addr().String(),
+			}, nil
+		}
+		return nil, nil
+	}
+	defer overwriteAndRestore(hpfe)()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	conn, err := grpc.Dial(unresolvedTargetURI, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("grpc.Dial failed: %v", err)
+	}
+	defer conn.Close()
+
+	client := testgrpc.NewTestServiceClient(conn)
+	if _, err := client.EmptyCall(ctx, &testgrpc.Empty{}); err != nil {
+		t.Errorf("EmptyCall failed: %v", err)
+	}
+
+	select {
+	case err := <-errCh:
+		t.Fatalf("proxy server encountered an error: %v", err)
+	case <-doneCh:
+		t.Logf("proxy server succeeded")
+	}
+}
+
+// TestGrpcDialWithProxyAndResolution tests grpc.Dial with a proxy and ensures DNS
+// resolution of the proxy URI is performed.
+func (s) TestGrpcDialWithProxyAndResolution(t *testing.T) {
+	// Create and start a backend server.
+	backendAddr := createAndStartBackendServer(t)
+
+	// Set up and start a proxy listener.
+	proxyLis, errCh, doneCh, _ := setupProxy(t, backendAddr, false, func(req *http.Request) error {
+		if req.Method != http.MethodConnect {
+			return fmt.Errorf("unexpected Method %q, want %q", req.Method, http.MethodConnect)
+		}
+		if req.URL.Host != unresolvedTargetURI {
+			return fmt.Errorf("unexpected URL.Host %q, want %q", req.URL.Host, unresolvedTargetURI)
+		}
+		return nil
+	})
+
+	// Overwrite the default proxy function and restore it after the test.
+	hpfe := func(req *http.Request) (*url.URL, error) {
+		if req.URL.Host == unresolvedTargetURI {
+			return &url.URL{
+				Scheme: "https",
+				Host:   unresolvedProxyURI,
+			}, nil
+		}
+		return nil, nil
+	}
+	defer overwriteAndRestore(hpfe)()
+
+	// Set up a manual resolver for proxy resolution.
+	mrProxy := setupDNS(t)
+
+	// Update the proxy resolver state with the proxy's address.
+	mrProxy.InitialState(resolver.State{
+		Addresses: []resolver.Address{
+			{Addr: proxyLis.Addr().String()},
+		},
+	})
+	// Dial to the proxy server.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	conn, err := grpc.Dial(unresolvedTargetURI, grpc.WithTransportCredentials(insecure.NewCredentials()))
```

Comment:
same comment about using `grpc.NewClient`

Reply:
Same : The test is specifically to test the behaviour of `grpc.Dial` , to check that it doesn't change. The other tests use `grpc.NewClient`